akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -0,0 +1,386 @@
+/**
+ *  Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.dispatch.affinity
+
+import java.lang.invoke.{ MethodHandle, MethodHandles }
+import java.lang.invoke.MethodType.methodType
+import java.util
+import java.util.Collections
+import java.util.concurrent.TimeUnit.MICROSECONDS
+import java.util.concurrent._
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.locks.{ Lock, LockSupport, ReentrantLock }
+
+import akka.dispatch._
+import akka.dispatch.affinity.AffinityPool._
+import akka.util.Helpers.Requiring
+import com.typesafe.config.Config
+
+import scala.annotation.tailrec
+import scala.collection.JavaConversions._
+import scala.util.Try
+import java.lang.Integer.reverseBytes
+
+import akka.util.OptionVal
+
+import scala.util.control.NonFatal
+
+private[akka] class AffinityPool(
+  parallelism:               Int,
+  affinityGroupSize:         Int,
+  tf:                        ThreadFactory,
+  idleCpuLevel:              Int,
+  fairDistributionThreshold: Int,
+  rejectionHandler:          RejectionHandler)
+  extends AbstractExecutorService {
+
+  if (parallelism <= 0)
+    throw new IllegalArgumentException("Size of pool cannot be less or equal to 0")
+
+  // Held while starting/shutting down workers/pool in order to make
+  // the operations linear and enforce atomicity. An example of that would be
+  // adding a worker. We want the creation of the worker, addition
+  // to the set and starting to worker to be an atomic action. Using
+  // a concurrent set would not give us that
+  private val bookKeepingLock = new ReentrantLock()
+
+  // condition used for awaiting termination
+  private val terminationCondition = bookKeepingLock.newCondition()
+
+  // indicates the current state of the pool
+  @volatile final private var poolState: PoolState = Running
+
+  private final val workQueues = Array.fill(parallelism)(new BoundedTaskQueue(affinityGroupSize))
+  private final val workers = new java.util.HashSet[ThreadPoolWorker]()
+
+  // a counter that gets incremented every time a task is queued
+  private val executionCounter: AtomicInteger = new AtomicInteger(0)
+  // maps a runnable to an index of a worker queue
+  private val runnableToWorkerQueueIndex: java.util.Map[Int, Int] = new ConcurrentHashMap[Int, Int]()
```

Comment:
I'm thinking of what we could do to make this more efficient:
* avoid boxing of `Int`
* concurrency overhead of ConcurrentHashMap

It would be rather simple to rewrite the https://github.com/akka/akka/blob/master/akka-remote/src/main/scala/akka/remote/artery/ImmutableLongMap.scala
to a specialized Int -> Int Map
Then that could be used with an `AtomicReference` and updated with `compareAndSet`.
The default size (before switching to pure hashCode) could be small, such as 128.